src/mongoose_cowboy_metrics_mw_after.erl

```diff
@@ -0,0 +1,121 @@
+%%
+%% This middleware does not create any metrics, they need to be created earlier using
+%% `mongoose_metrics' module. Metric names used by the middleware are contructed as follows:
+%%  - take some `Prefix', a list
+%%  - take a request method, `Method', one of:
+%%    - `<<"GET">>'
+%%    - `<<"HEAD">>'
+%%    - `<<"POST">>'
+%%    - `<<"PUT">>'
+%%    - `<<"DELETE">>'
+%%    - `<<"OPTIONS">>'
+%%    - `<<"PATCH">>'
+%% - take a request status class, `Class', and create a string like e.g. `<<"2XX">>' for success
+%%   status codes
+%% - for each `Method' and `Class' define following metric names
+%%   - `Prefix ++ [Method, request, count]' - updated by `1' whenever a request with method `Method'
+%%     is about to be handled
+%%   - `Prefix ++ [Method, response, Class, count]' - updated by `1' whenever a response of status
+%%      class `Class' to a request with method `Method' is sent
+%%   - `Prefix ++ [Method, response, Class, latency]' -  updated by number of microseconds which
+%%     passed since request timestamp was recorded by `mongoose_cowboy_metrics_mw_before' whenever
+%%     a response of status class `Class' to a request with method `Method' is sent
+%%
+%% As you might have already guessed it makes sense to define `count' metrics as spirals, and
+%% `latency' metrics as histograms. The middleware will always try to update the metric regardless
+%% of whether it was created. Note that it's run after `cowboy_router' middleware, which means that
+%% error responses returned by the router (such as 404 for no matching handler) won't be recorded.
+%%
+%% And what about `Prefix'? By default prefix is the name of the handler handling the
+%% request wrapped in a list. However, you might provide `handler_to_metric_prefix' map as Cowboy
+%% listener environment  value, where keys are handler names and values are corresponding prefixes.
+%%
+%% You can use functions from `mongoose_cowboy_metrics' module to generate names of metrics recorded
+%% by this module.
+%%
+%%==============================================================================
+
+-module(mongoose_cowboy_metrics_mw_after).
+
+-behaviour(cowboy_middleware).
+
+%% cowboy_middleware callbacks
+-export([execute/2]).
+
+%%-------------------------------------------------------------------
+%% cowboy_middleware callbacks
+%%-------------------------------------------------------------------
+
+execute(Req, Env) ->
+    case proplists:get_value(record_metrics, Env, false) of
+        true ->
+            {req_timestamp, StartTs} = proplists:lookup(req_timestamp, Env),
+            {handler, Handler} = proplists:lookup(handler, Env),
+            Method = get_req_method(Req),
+            HandlerToPrefixMappings = proplists:get_value(handler_to_metric_prefix, Env, #{}),
+            Prefix = maps:get(Handler, HandlerToPrefixMappings, [Handler]),
+            mongoose_metrics:update(global, mongoose_cowboy_metrics:request_count_metric(Prefix, Method), 1),
```

Comment:
Would it make sense to move the `request` hunter update to the `before` middleware?

Reply:
Unfortunately we can't do that 😢 "Before" middleware doesn't know which handler will serve the request, because `cowboy_router` has not been run yet.

On the other hand we can't merge the middleware, because in case of no matching routes cowboy_router terminates the middleware chain. In those cases we'd lose even the request count metric (currently we lose response count and latency).

Reply:
Ok, thanks for explanation :)

Reply:
@michalwski sorry, I was wrong. We can't report the metric until we know the handler, which means the reporting needs to occur after cowboy_router. However, if we're willing to **not** report the time which it takes for cowboy_router to execute (i.e. only measure the run time of the handler itself) then we could merge the middleware into one which could be placed after cowboy_router.